{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Testing model building with `carveme` and `memote`"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We firstly tried: Importing the COBRA module to read sbml models created via the `carveme` method. Subsequently we reconstruct a draft of the GSM model of *S. cerevisae* using the found RefSeq ID. Looking at the [reference genome R64](https://www.ncbi.nlm.nih.gov/data-hub/taxonomy/4932/?utm_source=assembly&utm_medium=referral&utm_campaign=KnownItemSensor:taxname), we see that 6464 have been annotated. \n",
    "\n",
    "But found a model online that worked better"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from cobra.io import read_sbml_model\n",
    "from cobra import Reaction, Metabolite\n",
    "import escher"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "%%capture\n",
    "!carve --refseq GCF_000146045.2_R64 --output Saccharomyces.xml"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "print(\"DONE WITH THE CARVING!\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "model = read_sbml_model('../yeast-GEM.xml') # Model initiated in minimal medium + glucose (flux bound = 1.0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "GSM model fact sheet:\n",
      "Number of reactions: 4058\n",
      "Number of metabolites: 2742\n",
      "Number of genes included in the model: 1150\n",
      "% of genes in model compared to the number of coding genes in Saccharomyces cerevisae R64 (6464 genes): 17.79\n",
      "Gene coverage percentage = 28.34 %\n"
     ]
    }
   ],
   "source": [
    "print(\"GSM model fact sheet:\")\n",
    "print(\"Number of reactions:\", len(model.reactions))\n",
    "print(\"Number of metabolites:\", len(model.metabolites))\n",
    "print(\"Number of genes included in the model:\", len(model.genes))\n",
    "print(\"% of genes in model compared to the number of coding genes in Saccharomyces cerevisae R64 (6464 genes):\",\n",
    "      round((len(model.genes)/6464)*100, 2))\n",
    "print(\"Gene coverage percentage =\", round(len(model.genes)/len(model.reactions)*100, 2), \"%\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Checking the model setup - how the metabolites etc. are named/tagged"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Metabolite s_0001[ce] at 0x7f03d8da7d30>,\n",
       " <Metabolite s_0002[c] at 0x7f03d8da7e50>,\n",
       " <Metabolite s_0003[e] at 0x7f03d8da7fa0>,\n",
       " <Metabolite s_0004[ce] at 0x7f03d8da7f40>,\n",
       " <Metabolite s_0006[m] at 0x7f03d8da7ee0>]"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites[:5]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'ce': 'cell envelope',\n",
       " 'c': 'cytoplasm',\n",
       " 'e': 'extracellular',\n",
       " 'm': 'mitochondrion',\n",
       " 'n': 'nucleus',\n",
       " 'p': 'peroxisome',\n",
       " 'er': 'endoplasmic reticulum',\n",
       " 'g': 'Golgi',\n",
       " 'lp': 'lipid particle',\n",
       " 'v': 'vacuole',\n",
       " 'erm': 'endoplasmic reticulum membrane',\n",
       " 'vm': 'vacuolar membrane',\n",
       " 'gm': 'Golgi membrane',\n",
       " 'mm': 'mitochondrial membrane'}"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.compartments # compartments used"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>s_0001[ce]</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>(1->3)-beta-D-glucan [cell envelope]</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f03d8da7d30</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C6H10O5</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>ce</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 3 reaction(s)</strong></td><td>\n",
       "                    r_1543, r_0005, r_4048</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite s_0001[ce] at 0x7f03d8da7d30>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.get_by_id('s_0001[ce]') # Again specifying model naming"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "acetoacetyl-CoA [cytoplasm]\n",
      "acetoacetyl-CoA [mitochondrion]\n",
      "acetyl-CoA [cytoplasm]\n",
      "acetyl-CoA [mitochondrion]\n",
      "acetyl-CoA [nucleus]\n",
      "acetyl-CoA [peroxisome]\n",
      "acetoacetyl-CoA [peroxisome]\n"
     ]
    }
   ],
   "source": [
    "# Investigating where in the model, acetyl-CoA is used\n",
    "for metabolite in model.metabolites.query('acetyl-CoA', 'name'):\n",
    "    print(metabolite.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Model objective function:\n",
      "Maximize\n",
      "1.0*r_2111 - 1.0*r_2111_reverse_58b69\n"
     ]
    }
   ],
   "source": [
    "# Model default objective function (growth rate)\n",
    "print(\"Model objective function:\\n\", model.objective, sep=\"\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_2111</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>growth</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f03d8c7d880</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>s_0450[c] --></p>\n",
       "                    <p style='text-align:right'>biomass [cytoplasm] --></p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_2111 at 0x7f03d8c7d880>"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Investigating the growth reaction rate\n",
    "model.reactions.get_by_id('r_2111')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>s_0450[c]</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>biomass [cytoplasm]</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f03d8d17f40</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>c</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 2 reaction(s)</strong></td><td>\n",
       "                    r_2111, r_4041</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite s_0450[c] at 0x7f03d8d17f40>"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.get_by_id('s_0450[c]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<strong><em>Optimal</em> solution with objective value 0.084</strong><br><div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fluxes</th>\n",
       "      <th>reduced_costs</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>r_0001</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.667286e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r_0002</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.667286e-02</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r_0003</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r_0004</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>1.734723e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r_0005</th>\n",
       "      <td>0.062686</td>\n",
       "      <td>6.722053e-18</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r_4694</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r_4695</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r_4697</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-0.000000e+00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r_4698</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>-1.866638e-01</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>r_4699</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000e+00</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>4058 rows × 2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "<Solution 0.084 at 0x7f03d84ff220>"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.optimize()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "r_1654 | ammonium exchange = 1000.0\n",
      "r_1714 | D-glucose exchange = 1.0\n",
      "r_1832 | H+ exchange = 1000.0\n",
      "r_1861 | iron(2+) exchange = 1000.0\n",
      "r_1992 | oxygen exchange = 1000.0\n",
      "r_2005 | phosphate exchange = 1000.0\n",
      "r_2020 | potassium exchange = 1000.0\n",
      "r_2049 | sodium exchange = 1000.0\n",
      "r_2060 | sulphate exchange = 1000.0\n",
      "r_2100 | water exchange = 1000.0\n",
      "r_4593 | chloride exchange = 1000.0\n",
      "r_4594 | Cu2(+) exchange = 1000.0\n",
      "r_4595 | Mn(2+) exchange = 1000.0\n",
      "r_4596 | Zn(2+) exchange = 1000.0\n",
      "r_4597 | Mg(2+) exchange = 1000.0\n",
      "r_4600 | Ca(2+) exchange = 1000.0\n"
     ]
    }
   ],
   "source": [
    "# Exchange reactions for uptake of nutrients from medium\n",
    "for r in model.medium.keys():\n",
    "    print(r, \"|\", model.reactions.get_by_id(r).name, \"=\",model.medium[r])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "r_1542 | (1->3)-beta-D-glucan exchange\n",
      "r_1545 | (R)-carnitine exchange\n",
      "r_1546 | (R)-lactate exchange\n",
      "r_1547 | (R)-mevalonate exchange\n",
      "r_1548 | (R)-pantothenate exchange\n",
      "r_1549 | (R,R)-2,3-butanediol exchange\n",
      "r_1550 | (S)-3-methyl-2-oxopentanoate exchange\n",
      "r_1551 | (S)-lactate exchange\n",
      "r_1552 | (S)-malate exchange\n",
      "r_1553 | 1-(sn-glycero-3-phospho)-1D-myo-inositol exchange\n",
      "r_1554 | 1-acylglycerophosphocholine exchange\n",
      "r_1563 | 2'-deoxyadenosine exchange\n",
      "r_1564 | 2'-deoxyguanosine exchange\n",
      "r_1565 | 2'-deoxyinosine exchange\n",
      "r_1566 | 2'-deoxyuridine exchange\n",
      "r_1572 | 2-isopropylmalate exchange\n",
      "r_1577 | 2-methylbutanal exchange\n",
      "r_1580 | 2-methylbutanol exchange\n",
      "r_1581 | 2-methylbutyl acetate exchange\n",
      "r_1586 | 2-oxoglutarate exchange\n",
      "r_1589 | 2-phenylethanol exchange\n",
      "r_1598 | 3-methylbutanal exchange\n",
      "r_1604 | 4-aminobenzoate exchange\n",
      "r_1621 | 5-aminolevulinate exchange\n",
      "r_1625 | 5-formyltetrahydrofolic acid exchange\n",
      "r_1627 | 7,8-diaminononanoate exchange\n",
      "r_1629 | 8-amino-7-oxononanoate exchange\n",
      "r_1630 | 9H-xanthine exchange\n",
      "r_1631 | acetaldehyde exchange\n",
      "r_1634 | acetate exchange\n",
      "r_1639 | adenine exchange\n",
      "r_1641 | adenosine 3',5'-bismonophosphate exchange\n",
      "r_1643 | adenosine exchange\n",
      "r_1648 | allantoate exchange\n",
      "r_1649 | allantoin exchange\n",
      "r_1650 | trehalose exchange\n",
      "r_1651 | alpha-D-glucosamine 6-phosphate exchange\n",
      "r_1654 | ammonium exchange\n",
      "r_1663 | bicarbonate exchange\n",
      "r_1671 | biotin exchange\n",
      "r_1672 | carbon dioxide exchange\n",
      "r_1683 | choline exchange\n",
      "r_1687 | citrate exchange\n",
      "r_1702 | cytidine exchange\n",
      "r_1705 | cytosine exchange\n",
      "r_1706 | D-arabinose exchange\n",
      "r_1709 | D-fructose exchange\n",
      "r_1710 | D-galactose exchange\n",
      "r_1711 | D-galacturonate exchange\n",
      "r_1712 | D-glucitol exchange\n",
      "r_1714 | D-glucose exchange\n",
      "r_1715 | D-mannose exchange\n",
      "r_1716 | D-ribose exchange\n",
      "r_1718 | D-xylose exchange\n",
      "r_1727 | decanoate exchange\n",
      "r_1730 | deoxycytidine exchange\n",
      "r_1749 | dTTP exchange\n",
      "r_1753 | episterol exchange\n",
      "r_1757 | ergosterol exchange\n",
      "r_1761 | ethanol exchange\n",
      "r_1764 | ethanolamine exchange\n",
      "r_1765 | ethyl acetate exchange\n",
      "r_1788 | fecosterol exchange\n",
      "r_1791 | FMN exchange\n",
      "r_1792 | folic acid exchange\n",
      "r_1793 | formate exchange\n",
      "r_1798 | fumarate exchange\n",
      "r_1800 | gamma-aminobutyrate exchange\n",
      "r_1806 | glutathione disulfide exchange\n",
      "r_1807 | glutathione exchange\n",
      "r_1808 | glycerol exchange\n",
      "r_1810 | L-glycine exchange\n",
      "r_1814 | glycolaldehyde exchange\n",
      "r_1815 | glyoxylate exchange\n",
      "r_1818 | guanine exchange\n",
      "r_1820 | guanosine exchange\n",
      "r_1832 | H+ exchange\n",
      "r_1834 | hexadecanal exchange\n",
      "r_1841 | hypoxanthine exchange\n",
      "r_1843 | indol-3-ylacetaldehyde exchange\n",
      "r_1847 | inosine exchange\n",
      "r_1861 | iron(2+) exchange\n",
      "r_1862 | isoamyl acetate exchange\n",
      "r_1865 | isoamylol exchange\n",
      "r_1866 | isobutanol exchange\n",
      "r_1867 | isobutyl acetate exchange\n",
      "r_1870 | isobutyraldehyde exchange\n",
      "r_1873 | L-alanine exchange\n",
      "r_1875 | L-arabinitol exchange\n",
      "r_1878 | L-arabinose exchange\n",
      "r_1879 | L-arginine exchange\n",
      "r_1880 | L-asparagine exchange\n",
      "r_1881 | L-aspartate exchange\n",
      "r_1883 | L-cysteine exchange\n",
      "r_1886 | L-glucitol exchange\n",
      "r_1889 | L-glutamate exchange\n",
      "r_1891 | L-glutamine exchange\n",
      "r_1893 | L-histidine exchange\n",
      "r_1896 | L-homoserine exchange\n",
      "r_1897 | L-isoleucine exchange\n",
      "r_1899 | L-leucine exchange\n",
      "r_1900 | L-lysine exchange\n",
      "r_1902 | L-methionine exchange\n",
      "r_1903 | L-phenylalanine exchange\n",
      "r_1904 | L-proline exchange\n",
      "r_1906 | L-serine exchange\n",
      "r_1909 | L-sorbose exchange\n",
      "r_1911 | L-threonine exchange\n",
      "r_1912 | L-tryptophan exchange\n",
      "r_1913 | L-tyrosine exchange\n",
      "r_1914 | L-valine exchange\n",
      "r_1915 | lanosterol exchange\n",
      "r_1916 | laurate exchange\n",
      "r_1931 | maltose exchange\n",
      "r_1947 | myo-inositol exchange\n",
      "r_1952 | N,N'-diformyldityrosine exchange\n",
      "r_1967 | nicotinate exchange\n",
      "r_1968 | NMN exchange\n",
      "r_1984 | octanoate exchange\n",
      "r_1987 | ornithine exchange\n",
      "r_1989 | oxaloacetate(2-) exchange\n",
      "r_1992 | oxygen exchange\n",
      "r_1993 | palmitate exchange\n",
      "r_1994 | palmitoleate exchange\n",
      "r_1999 | pectin exchange\n",
      "r_2000 | phenethyl acetate exchange\n",
      "r_2001 | phenylacetaldehyde exchange\n",
      "r_2005 | phosphate exchange\n",
      "r_2020 | potassium exchange\n",
      "r_2024 | putrescine exchange\n",
      "r_2028 | pyridoxine exchange\n",
      "r_2033 | pyruvate exchange\n",
      "r_2038 | riboflavin exchange\n",
      "r_2043 | S-adenosyl-L-methionine exchange\n",
      "r_2044 | S-methyl-L-methionine exchange\n",
      "r_2046 | sn-glycero-3-phosphocholine exchange\n",
      "r_2049 | sodium exchange\n",
      "r_2051 | spermidine exchange\n",
      "r_2052 | spermine exchange\n",
      "r_2055 | stearate exchange\n",
      "r_2056 | succinate exchange\n",
      "r_2058 | sucrose exchange\n",
      "r_2060 | sulphate exchange\n",
      "r_2061 | sulphite exchange\n",
      "r_2062 | taurine exchange\n",
      "r_2066 | thiamine(1+) diphosphate(1-) exchange\n",
      "r_2067 | thiamine(1+) exchange\n",
      "r_2068 | thiamine(1+) monophosphate exchange\n",
      "r_2073 | thymidine exchange\n",
      "r_2083 | tryptophol exchange\n",
      "r_2090 | uracil exchange\n",
      "r_2091 | urea exchange\n",
      "r_2092 | uridine exchange\n",
      "r_2100 | water exchange\n",
      "r_2102 | xanthosine exchange\n",
      "r_2104 | xylitol exchange\n",
      "r_2106 | zymosterol exchange\n",
      "r_2134 | 14-demethyllanosterol exchange\n",
      "r_2137 | ergosta-5,7,22,24(28)-tetraen-3beta-ol exchange\n",
      "r_2187 | butyrate exchange\n",
      "r_2188 | hexanoate exchange\n",
      "r_2189 | oleate exchange\n",
      "r_2193 | myristate exchange\n",
      "r_4043 | raffinose exchange\n",
      "r_4044 | melibiose exchange\n",
      "r_4442 | cysteamine exchange\n",
      "r_4494 | methanol exchange\n",
      "r_4496 | Ala-Gly exchange\n",
      "r_4497 | 3-oxalomalate(3-) exchange\n",
      "r_4498 | 6-O-alpha-D-glucopyranosyl-D-fructofuranose exchange\n",
      "r_4499 | 5-dehydro-D-gluconate exchange\n",
      "r_4500 | D-tagatose exchange\n",
      "r_4501 | turanose exchange\n",
      "r_4502 | D-glucose 6-phosphate exchange\n",
      "r_4503 | alpha-maltotriose exchange\n",
      "r_4504 | D-glucose 1-phosphate exchange\n",
      "r_4505 | methyl alpha-D-glucopyranoside exchange\n",
      "r_4506 | D-Glucosamine exchange\n",
      "r_4507 | glycerone exchange\n",
      "r_4508 | Ala-Gln exchange\n",
      "r_4509 | Ala-Leu exchange\n",
      "r_4510 | Gly-Gln exchange\n",
      "r_4511 | Ala-His exchange\n",
      "r_4512 | Gly-Asn exchange\n",
      "r_4513 | Ala-Glu exchange\n",
      "r_4514 | Gly-Met exchange\n",
      "r_4515 | Ala-Asp exchange\n",
      "r_4516 | Gly-Glu exchange\n",
      "r_4517 | Ala-Thr exchange\n",
      "r_4518 | Met-Ala exchange\n",
      "r_4519 | L-citrulline exchange\n",
      "r_4520 | 2-aminobutanoate exchange\n",
      "r_4521 | choline phosphate exchange\n",
      "r_4522 | glycerol 1-phosphate exchange\n",
      "r_4523 | 2-phospho-D-glyceric acid exchange\n",
      "r_4524 | GMP exchange\n",
      "r_4525 | 3-phospho-serine exchange\n",
      "r_4526 | O-phosphonatooxy-D-serine(2-) exchange\n",
      "r_4527 | diphosphate exchange\n",
      "r_4528 | triphosphate exchange\n",
      "r_4529 | phosphoenolpyruvate exchange\n",
      "r_4530 | N(omega)-phospho-L-arginine exchange\n",
      "r_4531 | carbamoyl phosphate exchange\n",
      "r_4532 | O-phosphoethanolamine exchange\n",
      "r_4533 | cysteamine S-phosphate exchange\n",
      "r_4534 | guanosine 2'-monophosphate exchange\n",
      "r_4535 | glycerol 2-phosphate(2-) exchange\n",
      "r_4536 | 3'-GMP exchange\n",
      "r_4537 | 2-phosphoglycolate exchange\n",
      "r_4538 | 6-phospho-D-gluconate exchange\n",
      "r_4539 | D-mannose 6-phosphate exchange\n",
      "r_4540 | 2',3'-cyclic GMP exchange\n",
      "r_4541 | O(4)-phospho-L-tyrosine exchange\n",
      "r_4542 | O-phospho-L-threonine exchange\n",
      "r_4543 | 3-phosphonato-D-glycerate(3-) exchange\n",
      "r_4544 | adenosine 2'-phosphate exchange\n",
      "r_4545 | 3'-AMP exchange\n",
      "r_4546 | 2',3'-cyclic AMP exchange\n",
      "r_4547 | D-mannose 1-phosphate exchange\n",
      "r_4548 | AMP exchange\n",
      "r_4549 | N(alpha)-acetyl-L-methionine exchange\n",
      "r_4550 | L-Methionine S-oxide exchange\n",
      "r_4551 | 3-sulfino-L-alanine exchange\n",
      "r_4552 | N-acetyl-L-cysteine exchange\n",
      "r_4553 | L-cysteate exchange\n",
      "r_4554 | 2-hydroxyethane-1-sulfonate exchange\n",
      "r_4555 | acetoacetate exchange\n",
      "r_4556 | N-acetyl-L-glutamate exchange\n",
      "r_4557 | UMP exchange\n",
      "r_4558 | CMP exchange\n",
      "r_4559 | thymidine 5'-monophosphate exchange\n",
      "r_4560 | thymidine 3'-monophosphate exchange\n",
      "r_4561 | cytidine 2'-phosphate exchange\n",
      "r_4562 | uridine 2'-phosphate exchange\n",
      "r_4563 | 3'-UMP exchange\n",
      "r_4564 | 2',3'-cyclic CMP exchange\n",
      "r_4565 | 2',3'-cyclic UMP exchange\n",
      "r_4593 | chloride exchange\n",
      "r_4594 | Cu2(+) exchange\n",
      "r_4595 | Mn(2+) exchange\n",
      "r_4596 | Zn(2+) exchange\n",
      "r_4597 | Mg(2+) exchange\n",
      "r_4600 | Ca(2+) exchange\n",
      "r_4644 | ethyl hexanoate exchange\n",
      "r_4645 | ethyl octanoate exchange\n",
      "r_4646 | ethyl butanoate exchange\n",
      "r_4647 | hexyl acetate exchange\n",
      "r_4648 | ethyl decanoate exchange\n",
      "r_4658 | propanal exchange\n",
      "r_4659 | propanol exchange\n",
      "r_4667 | methional exchange\n",
      "r_4668 | methionol exchange\n",
      "r_4676 | (4-hydroxyphenyl)acetaldehyde exchange\n",
      "r_4677 | tyrosol exchange\n",
      "r_4689 | tyrosyl acetate exchange\n",
      "r_4692 | methionyl acetate exchange\n",
      "r_4695 | propyl acetate exchange\n",
      "r_4699 | ethyl propionate exchange\n"
     ]
    }
   ],
   "source": [
    "# Available exchange reactions for our model\n",
    "for reaction in model.exchanges:\n",
    "    print(reaction.id, \"|\", reaction.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_1714</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>D-glucose exchange</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f03efe72310</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>s_0565[e] <=></p>\n",
       "                    <p style='text-align:right'>D-glucose [extracellular] <=></p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>-1.0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>1000.0</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_1714 at 0x7f03efe72310>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.r_1714"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Metabolite identifier</strong></td><td>s_0565[e]</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>D-glucose [extracellular]</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f03f156f1f0</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Formula</strong></td><td>C6H12O6</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Compartment</strong></td><td>e</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>In 6 reaction(s)</strong></td><td>\n",
       "                    r_1166, r_1714, r_0370, r_1024, r_4420, r_4400</td>\n",
       "            </tr>\n",
       "        </table>"
      ],
      "text/plain": [
       "<Metabolite s_0565[e] at 0x7f03f156f1f0>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.get_by_id('s_0565[e]')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "mu_min_gluc = 0.08374778664999834\n",
      "mu_max_gluc = 19.816830259464282\n"
     ]
    }
   ],
   "source": [
    "# Experimentation with glucose concentration --> reach maximum growth rate at conc. ca. 575 \n",
    "medium = model.medium\n",
    "min_gluc = model.optimize()\n",
    "print(\"mu_min_gluc =\", min_gluc.fluxes['r_2111'])\n",
    "with model:\n",
    "    medium['r_1714'] = 575  # Glucose exchange reaciton\n",
    "    model.medium = medium\n",
    "    solution = model.optimize()\n",
    "    print(\"mu_max_gluc =\",solution.fluxes['r_2111'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "acetyl-CoA C-acetyltransferase | r_0103: 2.0 s_0373[c] <=> s_0367[c] + s_0529[c]\n",
      "Acetoacetate:CoA ligase (AMP-forming) | r_4396: s_0434[c] + s_0529[c] + s_4132[c] --> s_0367[c] + s_0423[c] + s_0633[c]\n",
      "hydroxymethylglutaryl CoA synthase | r_0559: s_0367[c] + s_0373[c] + s_0803[c] --> s_0218[c] + s_0529[c] + s_0794[c]\n"
     ]
    }
   ],
   "source": [
    "for reaction in model.metabolites.get_by_id('s_0367[c]').reactions:\n",
    "    reaction\n",
    "    print(reaction.name, reaction, sep = \" | \")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "frozenset({<Reaction r_0103 at 0x7f03f0f53520>,\n",
       "           <Reaction r_0559 at 0x7f03f0c99c70>,\n",
       "           <Reaction r_4396 at 0x7f03d8893c70>})"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.metabolites.get_by_id('s_0367[c]').reactions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "r_4046 non-growth associated maintenance reaction\n"
     ]
    }
   ],
   "source": [
    "for r in model.reactions.query('maintenance', 'name'):\n",
    "    print(r.id, r.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "\n",
       "        <table>\n",
       "            <tr>\n",
       "                <td><strong>Reaction identifier</strong></td><td>r_4046</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Name</strong></td><td>non-growth associated maintenance reaction</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Memory address</strong></td>\n",
       "                <td>0x07f03d8ac4f70</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Stoichiometry</strong></td>\n",
       "                <td>\n",
       "                    <p style='text-align:right'>s_0434[c] + s_0803[c] --> s_0394[c] + s_0794[c] + s_1322[c]</p>\n",
       "                    <p style='text-align:right'>ATP [cytoplasm] + H2O [cytoplasm] --> ADP [cytoplasm] + H+ [cytoplasm] + phosphate [cytoplasm]</p>\n",
       "                </td>\n",
       "            </tr><tr>\n",
       "                <td><strong>GPR</strong></td><td></td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Lower bound</strong></td><td>0.7</td>\n",
       "            </tr><tr>\n",
       "                <td><strong>Upper bound</strong></td><td>0.7</td>\n",
       "            </tr>\n",
       "        </table>\n",
       "        "
      ],
      "text/plain": [
       "<Reaction r_4046 at 0x7f03d8ac4f70>"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.reactions.r_4046"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# code for changing objective function\n",
    "with model:\n",
    "    model.objective = model.reactions.'reaction # <-- here we insert the reaction we want as objective\n",
    "    print(model.optimize().objective_value)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "s_0189[c]\n",
      "geranylgeranyl diphosphate [cytoplasm]\n",
      "frozenset({<Reaction r_0461 at 0x7f03f0da4e80>, <Reaction r_0373 at 0x7f03f0da48e0>})\n",
      "s_0745[c]\n",
      "geranyl diphosphate [cytoplasm]\n",
      "frozenset({<Reaction r_0462 at 0x7f03f0d64eb0>, <Reaction r_0355 at 0x7f03f0d90fd0>})\n",
      "s_0746[lp]\n",
      "geranylgeranyl diphosphate [lipid particle]\n",
      "frozenset({<Reaction r_0282 at 0x7f03f0e83dc0>, <Reaction r_0281 at 0x7f03f0e72be0>})\n"
     ]
    }
   ],
   "source": [
    "for m in model.metabolites.query('geranyl', 'name'):\n",
    "    print(m.id, m.name, model.metabolites.get_by_id(str(m.id)).reactions, sep='\\n')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "hydroxymethylglutaryl CoA reductase \n",
      " r_0558: s_0218[c] + 2.0 s_0794[c] + 2.0 s_1212[c] --> s_0028[c] + s_0529[c] + 2.0 s_1207[c]\n",
      "hydroxymethylglutaryl-CoA transport \n",
      " r_1840: s_0218[c] <=> s_0221[m]\n",
      "hydroxymethylglutaryl CoA synthase \n",
      " r_0559: s_0367[c] + s_0373[c] + s_0803[c] --> s_0218[c] + s_0529[c] + s_0794[c]\n"
     ]
    }
   ],
   "source": [
    "for r in model.metabolites.get_by_id('s_0218[c]').reactions:\n",
    "    print(r.name, '\\n', r)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(1->3)-beta-D-glucan exchange\n",
      "(R)-carnitine exchange\n",
      "(R)-lactate exchange\n",
      "(R)-mevalonate exchange\n",
      "(R)-pantothenate exchange\n",
      "(R,R)-2,3-butanediol exchange\n",
      "(S)-3-methyl-2-oxopentanoate exchange\n",
      "(S)-lactate exchange\n",
      "(S)-malate exchange\n",
      "1-(sn-glycero-3-phospho)-1D-myo-inositol exchange\n",
      "1-acylglycerophosphocholine exchange\n",
      "2'-deoxyadenosine exchange\n",
      "2'-deoxyguanosine exchange\n",
      "2'-deoxyinosine exchange\n",
      "2'-deoxyuridine exchange\n",
      "2-isopropylmalate exchange\n",
      "2-methylbutanal exchange\n",
      "2-methylbutanol exchange\n",
      "2-methylbutyl acetate exchange\n",
      "2-oxoglutarate exchange\n",
      "2-phenylethanol exchange\n",
      "3-methylbutanal exchange\n",
      "4-aminobenzoate exchange\n",
      "5-aminolevulinate exchange\n",
      "5-formyltetrahydrofolic acid exchange\n",
      "7,8-diaminononanoate exchange\n",
      "8-amino-7-oxononanoate exchange\n",
      "9H-xanthine exchange\n",
      "acetaldehyde exchange\n",
      "acetate exchange\n",
      "adenine exchange\n",
      "adenosine 3',5'-bismonophosphate exchange\n",
      "adenosine exchange\n",
      "allantoate exchange\n",
      "allantoin exchange\n",
      "trehalose exchange\n",
      "alpha-D-glucosamine 6-phosphate exchange\n",
      "ammonium exchange\n",
      "bicarbonate exchange\n",
      "biotin exchange\n",
      "carbon dioxide exchange\n",
      "choline exchange\n",
      "citrate exchange\n",
      "cytidine exchange\n",
      "cytosine exchange\n",
      "D-arabinose exchange\n",
      "D-fructose exchange\n",
      "D-galactose exchange\n",
      "D-galacturonate exchange\n",
      "D-glucitol exchange\n",
      "D-glucose exchange\n",
      "D-mannose exchange\n",
      "D-ribose exchange\n",
      "D-xylose exchange\n",
      "decanoate exchange\n",
      "deoxycytidine exchange\n",
      "dTTP exchange\n",
      "episterol exchange\n",
      "ergosterol exchange\n",
      "ethanol exchange\n",
      "ethanolamine exchange\n",
      "ethyl acetate exchange\n",
      "fecosterol exchange\n",
      "FMN exchange\n",
      "folic acid exchange\n",
      "formate exchange\n",
      "fumarate exchange\n",
      "gamma-aminobutyrate exchange\n",
      "glutathione disulfide exchange\n",
      "glutathione exchange\n",
      "glycerol exchange\n",
      "L-glycine exchange\n",
      "glycolaldehyde exchange\n",
      "glyoxylate exchange\n",
      "guanine exchange\n",
      "guanosine exchange\n",
      "H+ exchange\n",
      "hexadecanal exchange\n",
      "hypoxanthine exchange\n",
      "indol-3-ylacetaldehyde exchange\n",
      "inosine exchange\n",
      "iron(2+) exchange\n",
      "isoamyl acetate exchange\n",
      "isoamylol exchange\n",
      "isobutanol exchange\n",
      "isobutyl acetate exchange\n",
      "isobutyraldehyde exchange\n",
      "L-alanine exchange\n",
      "L-arabinitol exchange\n",
      "L-arabinose exchange\n",
      "L-arginine exchange\n",
      "L-asparagine exchange\n",
      "L-aspartate exchange\n",
      "L-cysteine exchange\n",
      "L-glucitol exchange\n",
      "L-glutamate exchange\n",
      "L-glutamine exchange\n",
      "L-histidine exchange\n",
      "L-homoserine exchange\n",
      "L-isoleucine exchange\n",
      "L-leucine exchange\n",
      "L-lysine exchange\n",
      "L-methionine exchange\n",
      "L-phenylalanine exchange\n",
      "L-proline exchange\n",
      "L-serine exchange\n",
      "L-sorbose exchange\n",
      "L-threonine exchange\n",
      "L-tryptophan exchange\n",
      "L-tyrosine exchange\n",
      "L-valine exchange\n",
      "lanosterol exchange\n",
      "laurate exchange\n",
      "maltose exchange\n",
      "myo-inositol exchange\n",
      "N,N'-diformyldityrosine exchange\n",
      "nicotinate exchange\n",
      "NMN exchange\n",
      "octanoate exchange\n",
      "ornithine exchange\n",
      "oxaloacetate(2-) exchange\n",
      "oxygen exchange\n",
      "palmitate exchange\n",
      "palmitoleate exchange\n",
      "pectin exchange\n",
      "phenethyl acetate exchange\n",
      "phenylacetaldehyde exchange\n",
      "phosphate exchange\n",
      "potassium exchange\n",
      "putrescine exchange\n",
      "pyridoxine exchange\n",
      "pyruvate exchange\n",
      "riboflavin exchange\n",
      "S-adenosyl-L-methionine exchange\n",
      "S-methyl-L-methionine exchange\n",
      "sn-glycero-3-phosphocholine exchange\n",
      "sodium exchange\n",
      "spermidine exchange\n",
      "spermine exchange\n",
      "stearate exchange\n",
      "succinate exchange\n",
      "sucrose exchange\n",
      "sulphate exchange\n",
      "sulphite exchange\n",
      "taurine exchange\n",
      "thiamine(1+) diphosphate(1-) exchange\n",
      "thiamine(1+) exchange\n",
      "thiamine(1+) monophosphate exchange\n",
      "thymidine exchange\n",
      "tryptophol exchange\n",
      "uracil exchange\n",
      "urea exchange\n",
      "uridine exchange\n",
      "water exchange\n",
      "xanthosine exchange\n",
      "xylitol exchange\n",
      "zymosterol exchange\n",
      "growth\n",
      "14-demethyllanosterol exchange\n",
      "ergosta-5,7,22,24(28)-tetraen-3beta-ol exchange\n",
      "butyrate exchange\n",
      "hexanoate exchange\n",
      "oleate exchange\n",
      "myristate exchange\n",
      "raffinose exchange\n",
      "melibiose exchange\n",
      "lipid backbone exchange\n",
      "lipid chain exchange\n",
      "cysteamine exchange\n",
      "methanol exchange\n",
      "Ala-Gly exchange\n",
      "3-oxalomalate(3-) exchange\n",
      "6-O-alpha-D-glucopyranosyl-D-fructofuranose exchange\n",
      "5-dehydro-D-gluconate exchange\n",
      "D-tagatose exchange\n",
      "turanose exchange\n",
      "D-glucose 6-phosphate exchange\n",
      "alpha-maltotriose exchange\n",
      "D-glucose 1-phosphate exchange\n",
      "methyl alpha-D-glucopyranoside exchange\n",
      "D-Glucosamine exchange\n",
      "glycerone exchange\n",
      "Ala-Gln exchange\n",
      "Ala-Leu exchange\n",
      "Gly-Gln exchange\n",
      "Ala-His exchange\n",
      "Gly-Asn exchange\n",
      "Ala-Glu exchange\n",
      "Gly-Met exchange\n",
      "Ala-Asp exchange\n",
      "Gly-Glu exchange\n",
      "Ala-Thr exchange\n",
      "Met-Ala exchange\n",
      "L-citrulline exchange\n",
      "2-aminobutanoate exchange\n",
      "choline phosphate exchange\n",
      "glycerol 1-phosphate exchange\n",
      "2-phospho-D-glyceric acid exchange\n",
      "GMP exchange\n",
      "3-phospho-serine exchange\n",
      "O-phosphonatooxy-D-serine(2-) exchange\n",
      "diphosphate exchange\n",
      "triphosphate exchange\n",
      "phosphoenolpyruvate exchange\n",
      "N(omega)-phospho-L-arginine exchange\n",
      "carbamoyl phosphate exchange\n",
      "O-phosphoethanolamine exchange\n",
      "cysteamine S-phosphate exchange\n",
      "guanosine 2'-monophosphate exchange\n",
      "glycerol 2-phosphate(2-) exchange\n",
      "3'-GMP exchange\n",
      "2-phosphoglycolate exchange\n",
      "6-phospho-D-gluconate exchange\n",
      "D-mannose 6-phosphate exchange\n",
      "2',3'-cyclic GMP exchange\n",
      "O(4)-phospho-L-tyrosine exchange\n",
      "O-phospho-L-threonine exchange\n",
      "3-phosphonato-D-glycerate(3-) exchange\n",
      "adenosine 2'-phosphate exchange\n",
      "3'-AMP exchange\n",
      "2',3'-cyclic AMP exchange\n",
      "D-mannose 1-phosphate exchange\n",
      "AMP exchange\n",
      "N(alpha)-acetyl-L-methionine exchange\n",
      "L-Methionine S-oxide exchange\n",
      "3-sulfino-L-alanine exchange\n",
      "N-acetyl-L-cysteine exchange\n",
      "L-cysteate exchange\n",
      "2-hydroxyethane-1-sulfonate exchange\n",
      "acetoacetate exchange\n",
      "N-acetyl-L-glutamate exchange\n",
      "UMP exchange\n",
      "CMP exchange\n",
      "thymidine 5'-monophosphate exchange\n",
      "thymidine 3'-monophosphate exchange\n",
      "cytidine 2'-phosphate exchange\n",
      "uridine 2'-phosphate exchange\n",
      "3'-UMP exchange\n",
      "2',3'-cyclic CMP exchange\n",
      "2',3'-cyclic UMP exchange\n",
      "chloride exchange\n",
      "Cu2(+) exchange\n",
      "Mn(2+) exchange\n",
      "Zn(2+) exchange\n",
      "Mg(2+) exchange\n",
      "Ca(2+) exchange\n",
      "ethyl hexanoate exchange\n",
      "ethyl octanoate exchange\n",
      "ethyl butanoate exchange\n",
      "hexyl acetate exchange\n",
      "ethyl decanoate exchange\n",
      "propanal exchange\n",
      "propanol exchange\n",
      "methional exchange\n",
      "methionol exchange\n",
      "(4-hydroxyphenyl)acetaldehyde exchange\n",
      "tyrosol exchange\n",
      "tyrosyl acetate exchange\n",
      "methionyl acetate exchange\n",
      "propyl acetate exchange\n",
      "ethyl propionate exchange\n"
     ]
    }
   ],
   "source": [
    "for boundary in model.boundary:\n",
    "    print(boundary.name)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Reaction r_1542 at 0x7f03efee0700>,\n",
       " <Reaction r_1545 at 0x7f03efee0a90>,\n",
       " <Reaction r_1546 at 0x7f03efee0ca0>,\n",
       " <Reaction r_1547 at 0x7f03efee0f10>,\n",
       " <Reaction r_1548 at 0x7f03efee0e50>,\n",
       " <Reaction r_1549 at 0x7f03efef83d0>,\n",
       " <Reaction r_1550 at 0x7f03efef8550>,\n",
       " <Reaction r_1551 at 0x7f03efef8580>,\n",
       " <Reaction r_1552 at 0x7f03efef8640>,\n",
       " <Reaction r_1553 at 0x7f03efef8730>,\n",
       " <Reaction r_1554 at 0x7f03efef8790>,\n",
       " <Reaction r_1563 at 0x7f03efef8910>,\n",
       " <Reaction r_1564 at 0x7f03efef89a0>,\n",
       " <Reaction r_1565 at 0x7f03efef8a00>,\n",
       " <Reaction r_1566 at 0x7f03efef8a30>,\n",
       " <Reaction r_1572 at 0x7f03efef8bb0>,\n",
       " <Reaction r_1577 at 0x7f03effc2fd0>,\n",
       " <Reaction r_1580 at 0x7f03efef8d60>,\n",
       " <Reaction r_1581 at 0x7f03efef8e80>,\n",
       " <Reaction r_1586 at 0x7f03efef87c0>,\n",
       " <Reaction r_1589 at 0x7f03efef8dc0>,\n",
       " <Reaction r_1598 at 0x7f03efe92760>,\n",
       " <Reaction r_1604 at 0x7f03efef8b20>,\n",
       " <Reaction r_1621 at 0x7f03efe92dc0>,\n",
       " <Reaction r_1625 at 0x7f03efe92b80>,\n",
       " <Reaction r_1627 at 0x7f03efef8e20>,\n",
       " <Reaction r_1629 at 0x7f03efe92a90>,\n",
       " <Reaction r_1630 at 0x7f03efe92be0>,\n",
       " <Reaction r_1631 at 0x7f03efeacc70>,\n",
       " <Reaction r_1634 at 0x7f03efeacd60>,\n",
       " <Reaction r_1639 at 0x7f03efeac9d0>,\n",
       " <Reaction r_1641 at 0x7f03efeace80>,\n",
       " <Reaction r_1643 at 0x7f03efeaccd0>,\n",
       " <Reaction r_1648 at 0x7f03efeacd00>,\n",
       " <Reaction r_1649 at 0x7f03efeac880>,\n",
       " <Reaction r_1650 at 0x7f03efeacca0>,\n",
       " <Reaction r_1651 at 0x7f03efeacfa0>,\n",
       " <Reaction r_1654 at 0x7f03efeacee0>,\n",
       " <Reaction r_1663 at 0x7f03efec2940>,\n",
       " <Reaction r_1671 at 0x7f03efec2e50>,\n",
       " <Reaction r_1672 at 0x7f03efec2bb0>,\n",
       " <Reaction r_1683 at 0x7f03efec2fd0>,\n",
       " <Reaction r_1687 at 0x7f03efec28b0>,\n",
       " <Reaction r_1702 at 0x7f03efe5aca0>,\n",
       " <Reaction r_1705 at 0x7f03efe5aeb0>,\n",
       " <Reaction r_1706 at 0x7f03efe5aa00>,\n",
       " <Reaction r_1709 at 0x7f03efe5a790>,\n",
       " <Reaction r_1710 at 0x7f03efe5a850>,\n",
       " <Reaction r_1711 at 0x7f03efe5a760>,\n",
       " <Reaction r_1712 at 0x7f03efe5a550>,\n",
       " <Reaction r_1714 at 0x7f03efe72310>,\n",
       " <Reaction r_1715 at 0x7f03efe72640>,\n",
       " <Reaction r_1716 at 0x7f03efe72820>,\n",
       " <Reaction r_1718 at 0x7f03efe72940>,\n",
       " <Reaction r_1727 at 0x7f03efe72970>,\n",
       " <Reaction r_1730 at 0x7f03efe72df0>,\n",
       " <Reaction r_1749 at 0x7f03efe8b100>,\n",
       " <Reaction r_1753 at 0x7f03efe8b8b0>,\n",
       " <Reaction r_1757 at 0x7f03efe8be20>,\n",
       " <Reaction r_1761 at 0x7f03efe8bcd0>,\n",
       " <Reaction r_1764 at 0x7f03efe224c0>,\n",
       " <Reaction r_1765 at 0x7f03efe22640>,\n",
       " <Reaction r_1788 at 0x7f03efe22670>,\n",
       " <Reaction r_1791 at 0x7f03efe22d60>,\n",
       " <Reaction r_1792 at 0x7f03efe22940>,\n",
       " <Reaction r_1793 at 0x7f03efe22e50>,\n",
       " <Reaction r_1798 at 0x7f03efe22280>,\n",
       " <Reaction r_1800 at 0x7f03efe22fd0>,\n",
       " <Reaction r_1806 at 0x7f03efe22a60>,\n",
       " <Reaction r_1807 at 0x7f03efe22190>,\n",
       " <Reaction r_1808 at 0x7f03efe22880>,\n",
       " <Reaction r_1810 at 0x7f03efe3c520>,\n",
       " <Reaction r_1814 at 0x7f03efe3c9a0>,\n",
       " <Reaction r_1815 at 0x7f03efe3c070>,\n",
       " <Reaction r_1818 at 0x7f03efe3c880>,\n",
       " <Reaction r_1820 at 0x7f03efe3cd30>,\n",
       " <Reaction r_1832 at 0x7f03efe3cbe0>,\n",
       " <Reaction r_1834 at 0x7f03efe3ce20>,\n",
       " <Reaction r_1841 at 0x7f03efe3c490>,\n",
       " <Reaction r_1843 at 0x7f03efdd38e0>,\n",
       " <Reaction r_1847 at 0x7f03efdd3ac0>,\n",
       " <Reaction r_1861 at 0x7f03efdd3af0>,\n",
       " <Reaction r_1862 at 0x7f03efdd3dc0>,\n",
       " <Reaction r_1865 at 0x7f03efdd3d00>,\n",
       " <Reaction r_1866 at 0x7f03efdee5e0>,\n",
       " <Reaction r_1867 at 0x7f03efdee610>,\n",
       " <Reaction r_1870 at 0x7f03efdee760>,\n",
       " <Reaction r_1873 at 0x7f03efdee9d0>,\n",
       " <Reaction r_1875 at 0x7f03efdeeaf0>,\n",
       " <Reaction r_1878 at 0x7f03efdd3430>,\n",
       " <Reaction r_1879 at 0x7f03efdd36d0>,\n",
       " <Reaction r_1880 at 0x7f03efdd34f0>,\n",
       " <Reaction r_1881 at 0x7f03efdd35e0>,\n",
       " <Reaction r_1883 at 0x7f03efdeef40>,\n",
       " <Reaction r_1886 at 0x7f03efdeebe0>,\n",
       " <Reaction r_1889 at 0x7f03efdeed00>,\n",
       " <Reaction r_1891 at 0x7f03efdeee50>,\n",
       " <Reaction r_1893 at 0x7f03efdeefa0>,\n",
       " <Reaction r_1896 at 0x7f03efdee6d0>,\n",
       " <Reaction r_1897 at 0x7f03efe06340>,\n",
       " <Reaction r_1899 at 0x7f03efe065e0>,\n",
       " <Reaction r_1900 at 0x7f03efe06670>,\n",
       " <Reaction r_1902 at 0x7f03efe06700>,\n",
       " <Reaction r_1903 at 0x7f03efe067c0>,\n",
       " <Reaction r_1904 at 0x7f03efe06760>,\n",
       " <Reaction r_1906 at 0x7f03efe06880>,\n",
       " <Reaction r_1909 at 0x7f03efdeebb0>,\n",
       " <Reaction r_1911 at 0x7f03efdeec40>,\n",
       " <Reaction r_1912 at 0x7f03efdeecd0>,\n",
       " <Reaction r_1913 at 0x7f03efe06820>,\n",
       " <Reaction r_1914 at 0x7f03efe06ca0>,\n",
       " <Reaction r_1915 at 0x7f03efe06c40>,\n",
       " <Reaction r_1916 at 0x7f03efe06b20>,\n",
       " <Reaction r_1931 at 0x7f03efd9c910>,\n",
       " <Reaction r_1947 at 0x7f03efd9cd30>,\n",
       " <Reaction r_1952 at 0x7f03efd9ce50>,\n",
       " <Reaction r_1967 at 0x7f03efd9cf70>,\n",
       " <Reaction r_1968 at 0x7f03efd9c700>,\n",
       " <Reaction r_1984 at 0x7f03efdb8c70>,\n",
       " <Reaction r_1987 at 0x7f03efdb8d00>,\n",
       " <Reaction r_1989 at 0x7f03efdb8d90>,\n",
       " <Reaction r_1992 at 0x7f03efdb8a30>,\n",
       " <Reaction r_1993 at 0x7f03efdb8f70>,\n",
       " <Reaction r_1994 at 0x7f03efdb8f10>,\n",
       " <Reaction r_1999 at 0x7f03efdb8f40>,\n",
       " <Reaction r_2000 at 0x7f03efdb88b0>,\n",
       " <Reaction r_2001 at 0x7f03efdb8220>,\n",
       " <Reaction r_2005 at 0x7f03d8cb4670>,\n",
       " <Reaction r_2020 at 0x7f03efdb8e20>,\n",
       " <Reaction r_2024 at 0x7f03d8cb4250>,\n",
       " <Reaction r_2028 at 0x7f03d8cb4bb0>,\n",
       " <Reaction r_2033 at 0x7f03d8cb4fa0>,\n",
       " <Reaction r_2038 at 0x7f03d8cb47c0>,\n",
       " <Reaction r_2043 at 0x7f03d8cb4d30>,\n",
       " <Reaction r_2044 at 0x7f03d8cb4a60>,\n",
       " <Reaction r_2046 at 0x7f03d8cb4be0>,\n",
       " <Reaction r_2049 at 0x7f03d8cb4e50>,\n",
       " <Reaction r_2051 at 0x7f03d8cb4f10>,\n",
       " <Reaction r_2052 at 0x7f03d8ccb910>,\n",
       " <Reaction r_2055 at 0x7f03d8ccb9d0>,\n",
       " <Reaction r_2056 at 0x7f03d8ccba60>,\n",
       " <Reaction r_2058 at 0x7f03d8ccbaf0>,\n",
       " <Reaction r_2060 at 0x7f03d8ccbbb0>,\n",
       " <Reaction r_2061 at 0x7f03d8ccbb50>,\n",
       " <Reaction r_2062 at 0x7f03d8ccbc40>,\n",
       " <Reaction r_2066 at 0x7f03d8ccb5e0>,\n",
       " <Reaction r_2067 at 0x7f03d8ccbf10>,\n",
       " <Reaction r_2068 at 0x7f03d8ccbd90>,\n",
       " <Reaction r_2073 at 0x7f03d8ccbf40>,\n",
       " <Reaction r_2083 at 0x7f03d8ccbb20>,\n",
       " <Reaction r_2090 at 0x7f03d8ce4dc0>,\n",
       " <Reaction r_2091 at 0x7f03d8ce4e50>,\n",
       " <Reaction r_2092 at 0x7f03d8ce4eb0>,\n",
       " <Reaction r_2100 at 0x7f03d8ce47f0>,\n",
       " <Reaction r_2102 at 0x7f03d8ce4fd0>,\n",
       " <Reaction r_2104 at 0x7f03d8ce4730>,\n",
       " <Reaction r_2106 at 0x7f03d8ce4bb0>,\n",
       " <Reaction r_2134 at 0x7f03d8c7da00>,\n",
       " <Reaction r_2137 at 0x7f03d8c7d700>,\n",
       " <Reaction r_2187 at 0x7f03d8c7ddc0>,\n",
       " <Reaction r_2188 at 0x7f03d8c7de50>,\n",
       " <Reaction r_2189 at 0x7f03d8c7deb0>,\n",
       " <Reaction r_2193 at 0x7f03d8c7dfa0>,\n",
       " <Reaction r_4043 at 0x7f03d8ac4df0>,\n",
       " <Reaction r_4044 at 0x7f03d8ac4dc0>,\n",
       " <Reaction r_4442 at 0x7f03d8856f40>,\n",
       " <Reaction r_4494 at 0x7f03d8807d00>,\n",
       " <Reaction r_4496 at 0x7f03d881ec40>,\n",
       " <Reaction r_4497 at 0x7f03d881eeb0>,\n",
       " <Reaction r_4498 at 0x7f03d881ef40>,\n",
       " <Reaction r_4499 at 0x7f03d881ed60>,\n",
       " <Reaction r_4500 at 0x7f03d881ebe0>,\n",
       " <Reaction r_4501 at 0x7f03d881efd0>,\n",
       " <Reaction r_4502 at 0x7f03d881ef10>,\n",
       " <Reaction r_4503 at 0x7f03d881eb20>,\n",
       " <Reaction r_4504 at 0x7f03d881ea90>,\n",
       " <Reaction r_4505 at 0x7f03d881ee20>,\n",
       " <Reaction r_4506 at 0x7f03d881ee80>,\n",
       " <Reaction r_4507 at 0x7f03d881efa0>,\n",
       " <Reaction r_4508 at 0x7f03d881ec70>,\n",
       " <Reaction r_4509 at 0x7f03d881eb50>,\n",
       " <Reaction r_4510 at 0x7f03d881ed90>,\n",
       " <Reaction r_4511 at 0x7f03d87b60d0>,\n",
       " <Reaction r_4512 at 0x7f03d87b64c0>,\n",
       " <Reaction r_4513 at 0x7f03d87b6370>,\n",
       " <Reaction r_4514 at 0x7f03d87b6460>,\n",
       " <Reaction r_4515 at 0x7f03d87b6550>,\n",
       " <Reaction r_4516 at 0x7f03d87b65b0>,\n",
       " <Reaction r_4517 at 0x7f03d87b6610>,\n",
       " <Reaction r_4518 at 0x7f03d87b6670>,\n",
       " <Reaction r_4519 at 0x7f03d87b66d0>,\n",
       " <Reaction r_4520 at 0x7f03d87b6730>,\n",
       " <Reaction r_4521 at 0x7f03d87b6790>,\n",
       " <Reaction r_4522 at 0x7f03d87b67f0>,\n",
       " <Reaction r_4523 at 0x7f03d881e490>,\n",
       " <Reaction r_4524 at 0x7f03d87b6910>,\n",
       " <Reaction r_4525 at 0x7f03d87b69a0>,\n",
       " <Reaction r_4526 at 0x7f03d87b6970>,\n",
       " <Reaction r_4527 at 0x7f03d87b68e0>,\n",
       " <Reaction r_4528 at 0x7f03d87b6a30>,\n",
       " <Reaction r_4529 at 0x7f03d87b6a90>,\n",
       " <Reaction r_4530 at 0x7f03d87b68b0>,\n",
       " <Reaction r_4531 at 0x7f03d87b6be0>,\n",
       " <Reaction r_4532 at 0x7f03d87b6bb0>,\n",
       " <Reaction r_4533 at 0x7f03d87b63a0>,\n",
       " <Reaction r_4534 at 0x7f03d87b6d00>,\n",
       " <Reaction r_4535 at 0x7f03d87b6a60>,\n",
       " <Reaction r_4536 at 0x7f03d87b6af0>,\n",
       " <Reaction r_4537 at 0x7f03d87b6e20>,\n",
       " <Reaction r_4538 at 0x7f03d87b6c40>,\n",
       " <Reaction r_4539 at 0x7f03d87b6c70>,\n",
       " <Reaction r_4540 at 0x7f03d87b6eb0>,\n",
       " <Reaction r_4541 at 0x7f03d87b6f10>,\n",
       " <Reaction r_4542 at 0x7f03d87b6ca0>,\n",
       " <Reaction r_4543 at 0x7f03d87b6280>,\n",
       " <Reaction r_4544 at 0x7f03d87b6850>,\n",
       " <Reaction r_4545 at 0x7f03d87b6e50>,\n",
       " <Reaction r_4546 at 0x7f03d87b6dc0>,\n",
       " <Reaction r_4547 at 0x7f03d87b6f40>,\n",
       " <Reaction r_4548 at 0x7f03d87b61f0>,\n",
       " <Reaction r_4549 at 0x7f03d87b6130>,\n",
       " <Reaction r_4550 at 0x7f03d87b6190>,\n",
       " <Reaction r_4551 at 0x7f03d87b6fd0>,\n",
       " <Reaction r_4552 at 0x7f03d87b6ee0>,\n",
       " <Reaction r_4553 at 0x7f03d87b6cd0>,\n",
       " <Reaction r_4554 at 0x7f03d87d54c0>,\n",
       " <Reaction r_4555 at 0x7f03d87d5370>,\n",
       " <Reaction r_4556 at 0x7f03d87d5460>,\n",
       " <Reaction r_4557 at 0x7f03d87d5550>,\n",
       " <Reaction r_4558 at 0x7f03d87d55b0>,\n",
       " <Reaction r_4559 at 0x7f03d87d5610>,\n",
       " <Reaction r_4560 at 0x7f03d87d5670>,\n",
       " <Reaction r_4561 at 0x7f03d87d53a0>,\n",
       " <Reaction r_4562 at 0x7f03d87d5130>,\n",
       " <Reaction r_4563 at 0x7f03d87d56d0>,\n",
       " <Reaction r_4564 at 0x7f03d87d57c0>,\n",
       " <Reaction r_4565 at 0x7f03d87d58e0>,\n",
       " <Reaction r_4593 at 0x7f03d87d5820>,\n",
       " <Reaction r_4594 at 0x7f03d87d5880>,\n",
       " <Reaction r_4595 at 0x7f03d8770c70>,\n",
       " <Reaction r_4596 at 0x7f03d881e9a0>,\n",
       " <Reaction r_4597 at 0x7f03d8770f10>,\n",
       " <Reaction r_4600 at 0x7f03d8789d60>,\n",
       " <Reaction r_4644 at 0x7f03d873afa0>,\n",
       " <Reaction r_4645 at 0x7f03d873a730>,\n",
       " <Reaction r_4646 at 0x7f03d873afd0>,\n",
       " <Reaction r_4647 at 0x7f03d873a9d0>,\n",
       " <Reaction r_4648 at 0x7f03d873ad90>,\n",
       " <Reaction r_4658 at 0x7f03d873a970>,\n",
       " <Reaction r_4659 at 0x7f03d8753af0>,\n",
       " <Reaction r_4667 at 0x7f03d8753f10>,\n",
       " <Reaction r_4668 at 0x7f03d8753c10>,\n",
       " <Reaction r_4676 at 0x7f03d8753f40>,\n",
       " <Reaction r_4677 at 0x7f03d8768e80>,\n",
       " <Reaction r_4689 at 0x7f03d87537f0>,\n",
       " <Reaction r_4692 at 0x7f03d86fbc40>,\n",
       " <Reaction r_4695 at 0x7f03d86fb8b0>,\n",
       " <Reaction r_4699 at 0x7f03d86fbd00>]"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "model.exchanges"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
